delta-lake/delta-spark321db/src/main/scala/com/databricks/sql/transaction/tahoe/rapids/GpuOptimizeExecutor.scala

```diff
@@ -0,0 +1,401 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * This file was derived from:
+ *  1. DoAutoCompaction.scala from PR#1156 at https://github.com/delta-io/delta/pull/1156,
+ *  2. OptimizeTableCommand.scala from the Delta Lake project at https://github.com/delta-io/delta.
+ *
+ * Copyright (2021) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.databricks.sql.transaction.tahoe.rapids
+
+import java.util.ConcurrentModificationException
+
+import scala.annotation.tailrec
+import scala.collection.mutable.ArrayBuffer
+
+import com.databricks.sql.transaction.tahoe._
+import com.databricks.sql.transaction.tahoe.DeltaOperations.Operation
+import com.databricks.sql.transaction.tahoe.actions.{Action, AddFile, FileAction, RemoveFile}
+import com.databricks.sql.transaction.tahoe.commands.DeltaCommand
+import com.databricks.sql.transaction.tahoe.commands.optimize._
+import com.databricks.sql.transaction.tahoe.files.SQLMetricsReporting
+import com.databricks.sql.transaction.tahoe.sources.DeltaSQLConf
+import com.nvidia.spark.rapids.delta.RapidsDeltaSQLConf
+
+import org.apache.spark.SparkContext
+import org.apache.spark.SparkContext.SPARK_JOB_GROUP_ID
+import org.apache.spark.sql.{Row, SparkSession}
+import org.apache.spark.sql.catalyst.expressions.Expression
+import org.apache.spark.sql.execution.metric.SQLMetric
+import org.apache.spark.sql.execution.metric.SQLMetrics.createMetric
+import org.apache.spark.util.{SystemClock, ThreadUtils}
+
+class GpuOptimizeExecutor(
+                        sparkSession: SparkSession,
+                        txn: OptimisticTransaction,
+                        partitionPredicate: Seq[Expression],
+                        zOrderByColumns: Seq[String],
+                        prevCommitActions: Seq[Action])
+  extends DeltaCommand with SQLMetricsReporting with Serializable {
+
+  /** Timestamp to use in [[FileAction]] */
+  private val operationTimestamp = new SystemClock().getTimeMillis()
```

Comment:
minor nit: we can remove the use of this Spark class and just call `System.getTimeMillis()` (which is all this does).

```suggestion
  private val operationTimestamp = System.getTimeMillis()
```

Reply:
Another question here. This timestamp is initialized once when the class is created and then re-used for all future operations. Is that intended?

Reply:
I'll change to `System.getTimeMillis()`, like you suggest. Thank you.

> This timestamp is initialized once when the class is created and then re-used for all future operations. Is that intended?

The intention (AFAICT) is to use the timestamp (from the creation of the executor) for all `FileAction`s issued in the executor.
My understanding is that there will be a different timestamp for each `GpuOptimizeExecutor` instance. Did I get that wrong?

Reply:
Yes, each `GpuOptimizeExecutor` will get its own timestamp. All actions performed by a `GpuOptimizeExecutor` will use this constraint timestamp. If that is the intention, then LGTM.